### DISCUSSION TITLE
Differentiable rigid-fluid coupling suggestion

### DISCUSSION THREAD

**Discussion by BinglunWang** [Genesis-Embodied-AI/genesis-world#405] [Ideas]
We could integreate DiffFR: Differentiable SPH-based Fluid-Rigid Coupling for Rigid Body Control [ToG2023, Li et al.]

https://zhehaoli1999.github.io/DiffFR/

**Comment by zhouxian** [comment]
I am not sure how different it is than the sph pipeline we have for now (haven't read the website in detail)

We have the whole differentiable framework ready, just need to add manual gradient checkpointing and computation graph for both rigid and sph -- is what you are suggesting more different? if so, could you briefly explain? thank you!!

**Reply by BinglunWang** [reply]
Thank you for your reply.

autodiff SPH might not be stable. Here's some disadvantages of autodiff SPH in my current understanding:
1. Autodiff rigid-fluid coupling can led to memory & speed issue. You can assume force is constantly transmitted in particles;
2. SPH kernel function second derivative can have the exploding gradients problem;

And in diffFR, they present autodiff have the above issue 2 & 3. And their main method is cut-off the gradients out of rigid body neighborhood and ignore 2nd derivative of SPH kernel function. So diffFR, it make DFSPH differentiable and ignore some computing to mitigate the issue 2 & 3.

**Reply by zhouxian** [reply]
I see, good to know! In our implementation, autodiff (plus manual gradient ckpt) through up to thousands of timesteps is still relatively stable even for MPM fluid. SPH might be a bit different, we will see.

If you have spare effort to implement this, it would also be greatly appreciated! :)

**Reply by BinglunWang** [reply]
> I see, good to know! In our implementation, autodiff (plus manual gradient ckpt) through up to thousands of timesteps is still relatively stable even for MPM fluid. SPH might be a bit different, we will see.
> 
> 
> 
> If you have spare effort to implement this, it would also be greatly appreciated! :)
> 
> 
> 
> 

Thanks. Sure, happy to contribute. If I have chances will implement it and let you know. 

And may I know is it released now for rigid-fluid(MPM) method? Or when you plan to release it?

And is it supported for differentiable **articulated** rigid bodys-fluid systems?

**Reply by zhouxian** [reply]
Mpm diff is released. There’s a material called tool, which is a temporary workaround for rigid material and supports differentiable coupling

Diff articulated rigid will be released soon. In Feb or so
